### PR TITLE
feat(ipos): filebase IPFS worker 

### LINKED ISSUE #304: R2(IPFS) worker
### Context 

[nft.storage has end of life in 30/06/204](https://nft.storage/blog/important-product-updates-nft-storage-is-evolving) therefore [nft-storage worker wont work](https://github.com/kodadot/workers/tree/main/nft-storage).

Therefore as I was speaking on last dev call we need a replacement for it :)

### Task

1. Create new worker called `ipos` (stands for inter-planetary object storage)
2. make same API as `nft-storage` worker (`/pinJson`, `/pinFile`)
3. the content should be uploaded to R2 under the "created" CID
4. Ideally if same response is used

### Tech stack 

Hono + R2 + (aaand maybe KV?)

Library wise you would probably need https://github.com/alanshaw/ipfs-only-hash
**or**
https://github.com/ipfs/js-ipfs-unixfs/tree/main/packages/ipfs-unixfs-importer

with 
- https://github.com/ipfs/js-stores/tree/main/packages/datastore-s3
- or https://github.com/ipfs/js-stores/tree/main/packages/blockstore-s3

### Routes

Screenshot 2024-06-14 at 15 50 57

Screenshot 2024-06-14 at 15 51 05

Screenshot 2024-06-14 at 15 51 11

### PR DESCRIPTION
**Pull request by hassnian**
## Context

add new `ipos` worker

### envs 

bucket: https://docs.filebase.com/getting-started/ipfs-getting-started-guide#how-to-create-an-ipfs-bucket
s3: https://docs.filebase.com/getting-started/s3-api-getting-started-guide#access-keys

````
 // S3
  S3_ACCESS_KEY_ID: string
  S3_SECRET_ACCESS_KEY: string //  secret env

  // Filebase
  FILEBASE_BUCKET_NAME: string 
````

## Ref

- [x] Closes #304
- [ ] References kodadot/<indexer/worker>#<issue_number>

## PR Type

- [ ] Bugfix
- [x] Feature
- [ ] Refactoring
- [ ] Chore

### Screenshots

CleanShot 2024-07-17 at 13 48 09@2x

CleanShot 2024-07-17 at 13 48 35@2x

### CODE REVIEW

**Comment by hassnian** [comment]
core packages of the propposed stack are not yet compatible with cloudlflare worker's environment

do we change env or go with another alternative cc @vikiival @preschian 

https://github.com/kodadot/workers/issues/304

CleanShot 2024-07-15 at 12 06 06@2x


mainly 
- https://www.npmjs.com/package/@libp2p/tcp

ref:
- https://community.cloudflare.com/t/how-to-make-node-built-in-modules-work/486465
- https://community.cloudflare.com/t/wrangler-polyfill-nodejs-events-not-working/638259?__cf_chl_rt_tk=tTvo5V2L0F7kT1rghwvy3IqoT3WZKPqqPDPqYAsVR3s-1720863852-0.0.1.1-6356
- net module

CleanShot 2024-07-15 at 12 09 40@2x

**Comment by vikiival** [comment]
Hey,
Thanks for taking it. 

We will use filebase. 
I will make new issue for that

File: services/ipos/src/routes/pinning.ts [hunk 1]
```diff
@@ -0,0 +1,120 @@
...
+  const cid = await Hash.of(content)
+
+  await s3.putObject({
+    Body: content,
+    Bucket: c.env.FILEBASE_BUCKET_NAME,
+    Key: cid,
+    ContentType: type,
+  })
+
+  await c.env.BUCKET.put(cid, new Blob([content], { type }))
```
Comment (vikiival) [inline services/ipos/src/routes/pinning.ts:30]: you can use wait until so it's non blocking op

**Comment by hassnian** [comment]
@vikiival @preschian pls create a new worker called `ipos` and update `wrangler.tioml` , so we can test this out 

thanks

**Comment by vikiival** [comment]
@SocketSecurity ignore aws-sdk@2.1659.0

**Comment by vikiival** [comment]
Tested, works as charm

**Comment by vikiival** [comment]
@SocketSecurity ignore npm/aws-sdk@2.1659.0

**Review by vikiival — APPROVED** [review]

**Pull request merged**